library/src/main/java/com/datatorrent/lib/iteration/nDelayOperator.java

```diff
@@ -0,0 +1,186 @@
+
+/**
+ * This operator is used in iteration and can do n window delay.
+ * Required delay needs to be set in the constructor, default delay is 1.
+  *
+ * @displayName nDelayOperator
+ * @category iteration
+
+ */
+public class nDelayOperator<T> implements Operator.DelayOperator, Operator.CheckpointListener
+{
+  private WindowDataManager windowDataManager = new WindowDataManager.FSWindowDataManager();
+  @Min(1)
+  private int delay = 1;
+  private long currentWindowId;
+  private transient int operatorContextId;
+  private transient ArrayList<T> windowData;
+  private transient Context.OperatorContext context;
+
+  public transient DefaultInputPort<T> input = new DefaultInputPort<T>() {
+    @Override
+    public void process(T t)
+    {
+      processTuple(t);
+    }
+  };
+
+  public transient DefaultOutputPort<T> output = new DefaultOutputPort();
+
+  public nDelayOperator()
+  {
+    init();
+  }
+
+  /*
+  * @param delay set the delay in number of windows for the tuples.
+   */
+  public nDelayOperator(int delay)
+  {
+    if ( delay < 1 ) {
+      throw new IllegalArgumentException("Invalid Delay specified.");
+    }
+    this.delay = delay;
+    init();
+  }
+
+  /*
+* Get Window Data Manager instance
+*/
+  public WindowDataManager getWindowDataManager()
+  {
+    return windowDataManager;
+  }
+
+  /*
+  * Set Window Data Manager instance
+  */
+  public void setWindowDataManager(WindowDataManager windowDataManager)
+  {
+    this.windowDataManager = windowDataManager;
+  }
+
+  private void init()
+  {
+    windowData = new ArrayList<>();
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    this.operatorContextId = context.getId();
+    this.windowDataManager.setup(context);
+    this.context = context;
+  }
+
+  @Override
+  public void teardown()
+  {
+    this.windowDataManager.teardown();
+  }
+
+  @Override
+  public void firstWindow()
+  {
+    replay(currentWindowId - delay);
+  }
+
+  private void replay( long windowId )
+  {
+    if ( windowId < 0 ) {
+      return;
+    }
+
+    ArrayList<T> recoveredData;
+    try {
+      recoveredData = (ArrayList<T>)this.windowDataManager.load(operatorContextId, windowId);
+      if (recoveredData == null) {
+        return;
+      }
+      for ( T tuple : recoveredData) {
+        output.emit(tuple);
+      }
+    } catch (IOException e) {
+      DTThrowable.rethrow(e);
+    }
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    currentWindowId = windowId;
+
+    if ( delay > 1 ) {
+      replay(windowId - delay + 1);
+    }
+  }
+
+  @Override
+  public void endWindow()
+  {
+    try {
+      this.windowDataManager.save(windowData, operatorContextId, currentWindowId);
+    } catch (IOException e) {
+      DTThrowable.rethrow(e);
+    }
+
+    windowData.clear();
+  }
+
+  protected void processTuple(T tuple)
+  {
+    windowData.add(tuple);
+
+    if ( delay == 1 ) {
+      output.emit(tuple);
+    }
```

Comment:
Tuples are just stored in WindowDataManager and replayed during BeginWindow. So it is important not to change delay after the object is constructed, so it can only be set through constructor.